smolyay/surrogate.py

```diff
@@ -315,3 +369,128 @@ def _reset_grid(self):
+    grid_generator: IndexGridGenerator
+        Set of grids and their corresponding basis function.
+
+    Raises
+    ------
+    TypeError
+        Grids must be :class:`IndexGridGenerator`.
+        Domain must be dim x 2.
+
+    Example
+    -------
+    Consider a function :math:F(y) with one independent variable.
+    Basis function B_n(x) is chosen for the approximation and
+    three grid points and their corresponding basis functions
+    are generated in (-1, 1) domain:
+    (:math:x_0, :math:x_1, :math:x_2)
+    [:math:B_0(x), :math:B_1(x) :math:B_2(x)]
+    Based on the domain of the approximation, grid points of the basis
+    functions can be moved to the new domain, resulting in transformed
+    grid points:
+    ..math:
+        (y_0, y_1, y_2)
+    Then:
+    ..math::
+        \[
+        \begin{bmatrix}
+        B_0'(x_0)& B_1'(x_0) & B_2'(x_0) \\
+        B_0'(x_1)& B_1'(x_1) & B_2'(x_1) \\
+        B_0'(x_2)& B_1'(x_2) & B_2'(x_2)
+        \end{bmatrix}
+        \begin{bmatrix}
+        C_0\\C_1\\C_2
+        \end{bmatrix}
+        =\begin{bmatrix}
+        F(y_0)\\F(y_1)\\F(y_2)
+        \end{bmatrix}
+        \]
+
+    Coefficients can be obtained by solving Ax=B
+    (overdetermined equations for dimension > 1).
+    ..math:
+        S = C_0B_0(x) + C_1B_1(x) + C_2B_2(x)
+    """
+
+    def __init__(self, domain, grid_generator):
+        super().__init__(domain, grid_generator)
+
+    def train(self, gradient_function):
+        """Fit surrogate's components to the gradient of a function.
+
+        Parameters
+        ----------
+        gradient_function: callabe
+            Gradient of a function to be approximated.
+        """
+        data = [gradient_function(point) for point in self.points]
+        self.train_from_data(data)
+
+    def train_from_data(self, data):
+        """Fit surrogate's components to data (gradient of function).
+
+        Parameters
+        ----------
+        data: list
+            Gradient function at grid points.
+
+        Raises
+        ------
+        IndexError
+            Data should be of length of the grid points x dimension.
+        """
+        self._data = numpy.array(data, ndmin=1)
+        expected_data_shape = ((len(self.grid.points), self.dimension)
+                               if self.dimension > 1 else (
+                                       len(self.grid.points),))
+        if self._data.shape != expected_data_shape:
+            raise IndexError("Data must be of size of number"
+                             "of grid points x dimension.")
+        points, basis_functions = self.grid.points, self.grid.basis_functions
+        # create basis matrix with appropriate size
+        basis_matrix = numpy.zeros((self.dimension * len(points), len(points)))
+        # evalaute each term
+        count = -1
+        for i, point in enumerate(points):
+            for ni in range(self.dimension):
+                count += 1
```

Comment:
I would move this below and put after you assign the value to `basis_matrix`. Then, you can start `count = 0`. This would be a more typical pattern.